net/device/IotHub.cs

```diff
@@ -75,10 +87,97 @@ public bool Cleanup(bool force = false)
             return false;
         }
 
-        public async Task SendTelemetryAsync(
-            object telemetryObject,
-            IDictionary<string, string> extraProperties,
-            CancellationToken cancellationToken)
+        public async Task RunAsync(CancellationToken ct)
+        {
+            while (!ct.IsCancellationRequested)
+            {
+                // Wait when there are few messages to send, or if not connected
+                if (_messagesToSend.Count < 10
+                    || !_isConnected)
+                {
+                    try
+                    {
+                        await Task.Delay(60000, ct).ConfigureAwait(false);
+                    }
+                    catch (TaskCanceledException)
+                    {
+                        // App is signalled to exit
+                        return;
+                    }
+                }
+
+                _logger.Metric("MessageBacklog", _messagesToSend.Count);
+
+                // If not connected, skip the work below this round
+                if (!_isConnected)
+                {
+                    _logger.Trace($"Waiting for connection before batching telemetry", TraceSeverity.Warning);
+                    continue;
+                }
+
+                // Make a batch of messages to send, unless we're retrying a previous batch
+                if (!_pendingMessages.Any()
+                    && _messagesToSend.Any())
```

Comment:
In case of long periods of disconnect, `_messagesToSend` will keep on getting bigger, without any message being actually sent out. Should we be worrying about storage space constraints?

Reply:
I'm not worrying about memory constraints until I find we have to. I concur it is a potential problem though.